external/llvm-project/mlir/lib/Transforms/Utils/OutlinerUtils.cpp

```diff
@@ -173,337 +209,380 @@ static bool opsMatch(Operation *lhs, Operation *rhs, OutliningCandidate &one,
+    // function. ? Look for earliest result ?
+    b.setInsertionPointAfter(_leadingOps.back());
+    func::CallOp callOp =
+        b.create<func::CallOp>(fusedLoc, outlinedFunc, _inputs.getArrayRef());
 
-  // Erase the ops we outlined, which should be safe now.
-  for (auto &op : llvm::make_early_inc_range(llvm::reverse(trailingOps))) {
-    if (op->use_empty()) {
-      op->erase();
-    }
-  }
-  assert(anchorOp->use_empty() && "expected 'op' to have no uses");
-  anchorOp->erase();
-  for (auto &op : llvm::make_early_inc_range(leadingOps)) {
-    if (op->use_empty()) {
-      op->erase();
+    for (auto it : llvm::zip(_results, callOp->getResults())) {
+      std::get<0>(it).replaceAllUsesWith(std::get<1>(it));
     }
   }
-}
 
-void Outliner::traceInputs(Operation *op, Operation *ignoreOp,
-                           SetVector<Operation *> &predecessors,
-                           SetVector<Value> &inputNodes) {
-  for (const auto &opnd : op->getOperands()) {
-    Operation *usedOp = opnd.getDefiningOp();
-    if (usedOp == ignoreOp)
-      continue;
-    if (usedOp && isLeadingOp(usedOp)) {
-      if (predecessors.contains(
-              usedOp)) // If already present, move it for new use.
-        predecessors.remove(usedOp);
-      predecessors.insert(usedOp);
-      if (!usedOp->hasTrait<OpTrait::ConstantLike>()) {
-        // depth first
-        traceInputs(usedOp, op, predecessors, inputNodes);
-      }
-    } else if (!predecessors.contains(
-                   usedOp)) { // special-case consts aren't inputs
-      inputNodes.insert(opnd);
-    }
-  }
-}
+private:
+  Outliner &_outliner;
+
+  Operation *_anchorOp;
+  Block *_anchorBlock;
+
+  SmallVector<Location, FUSION_CAPACITY_DEFAULT> _locs;
 
-// Inspired by / adapted from TestSCFIfUtilsPass in
-// test/lib/Transforms/TestSCFUtils.cpp.
+  typedef SmallVector<Value, FUSION_CAPACITY_DEFAULT> ValVec;
+  typedef SmallVector<Operation *, FUSION_CAPACITY_DEFAULT> OpVec;
+  SetVector<Value, ValVec> _inputs;
+  SetVector<Operation *, OpVec> _leadingOps;
+  SetVector<Operation *, OpVec> _trailingOps;
+  SetVector<Value, ValVec> _results;
+};
+
+// Walk each func outlining fusion opportunities, replacing with a call
+// to a new (or matching) function containing the functionality. The call
+// will be annotated the loc's of all fused ops.
 void Outliner::outline(ModuleOp module) {
   auto funcOps = module.getOps<func::FuncOp>();
+
   for (auto func : llvm::make_early_inc_range(funcOps)) {
     // Don't outline a kernel;  it may already have been outlined.
     if (func->hasAttr(outlineTag))
       continue;
 
+    bool hasMemrefs = false;
     std::vector<Operation *> anchors;
     auto callback = [&](Operation *op) {
+      for (auto operand : op->getOperands()) {
+        if (isa<MemRefType>(operand.getType()))
+          hasMemrefs = true;
+      }
+      for (auto result : op->getResults()) {
+        if (isa<MemRefType>(result.getType()))
+          hasMemrefs = true;
+      }
```

Comment:
Will be some redundant touches here, since (a) most results become somebody's operand and (b) we keep looking after hasMemrefs is true.